pkg/lib/bundle/generate.go

```diff
@@ -46,30 +50,27 @@ type AnnotationMetadata struct {
 // @channels: The list of channels that bundle image belongs to
 // @channelDefault: The default channel for the bundle image
 // @overwrite: Boolean flag to enable overwriting annotations.yaml locally if existed
-func GenerateFunc(directory, outputDir, packageName, channels, channelDefault string, overwrite bool) error {
-	// clean the input so that we know the absolute paths of input directories
-	directory, err := filepath.Abs(directory)
-	if err != nil {
-		return err
-	}
-	if outputDir != "" {
-		outputDir, err = filepath.Abs(outputDir)
-		if err != nil {
-			return err
-		}
-	}
-
-	_, err = os.Stat(directory)
-	if os.IsNotExist(err) {
+func GenerateFunc(option ...BundleOption) error {
+	bundleConfig := &BundleConfig{}
+	bundleConfig.apply(option)
+	if err := bundleConfig.complete(); err != nil {
 		return err
 	}
 
 	// Determine mediaType
-	mediaType, err := GetMediaType(directory)
+	mediaType, err := GetMediaType(bundleConfig.bundleDir)
 	if err != nil {
 		return err
 	}
 
+	if mediaType == PlainType {
+		if err = synthesize.Generate(path.Join(bundleConfig.bundleDir, MetadataDir), ManifestsDir, bundleConfig.outputDir, bundleConfig.overwrite); err != nil {
```

Comment:
From what I can tell, the behavior of `generate` depends on a few variables:

1.  the olm.yaml file; holds non-inferable metadata
2. the dependencies.yaml file; holds non-inferable metadata (required APIs)
3. extra k8s manifests, not normally in a bundle; the source of inferable metadata
4  the existence of a CSV; when absent tells us to synthesize one
5. the existence of an annotations.yaml file; tells us where to find olm.yaml, dependencies.yaml, itself, and the k8s resource; when absent tells us to synthesize one
6. the `overwrite` field

I'm having a hard time figuring out which combinations of these variables are valid. @Bowenislandsong could you explain the valid input and expected output, and maybe the use cases behind each?

Reply:
That's a very valid question. It was confusing but I thought it should be a good user experience because of the embedded inference for the following scenarios (guarded by `TestGenerateCSVlessBundle` test where @ecordell pointed out that I am not comparing the outcome but the point of the test is for this purpose https://github.com/operator-framework/operator-registry/pull/252#discussion_r424053619): 
1. A bundle that has no CSV but has non-inferable in metadata/
  - generate CSV
2. A bundle that has CSV and non-inferable in metadata/
 - does not generate CSV unless --overwrite
3. A bundle that has no CSV and no non-inferable in metadata/ 
 - return error that a bundle has to have at least either CSV or non-inferable info
4. A bundle that has CSV and no non-inferable in metadata/ 
 - does nothing about generating CSV

Reply:
I guess it was the initial intention there. I guess I forgot to update the test.

Reply:
for others, if there is no dependencies.yaml, we do not have required apis, if there is no annotations.yaml, we have metadata/ as metadata (this is not the case anymore because dependency work hard coded metadata dir.)

Reply:
Please let me know if that answers your question.

Reply:
I have updated my test to cover all situations.

Reply:
https://github.com/Bowenislandsong/operator-registry/blob/80a58b21243a7e66fb44478e8aed89ec2a400e9e/pkg/lib/bundle/generate_test.go#L312

Reply:
outdated as we divorced generate and generate csv.